### PR TITLE
Al gaussian process

### PR DESCRIPTION
**Pull request by feeds**

### CODE REVIEW

File: active_learning/active_learner.py [hunk 1]
```diff
@@ -1,68 +1,185 @@
...
+
+    def _prob_choose_meta_path(self) -> np.ndarray:
+
+        if sum(self.visited):
+            return self.visited / sum(self.visited)
+        else:
+            return np.append(np.zeros(len(self.meta_paths) - 1), [1])
+
+
+class UncertaintySamplingAlgorithm(ActiveLearningAlgorithm):
```
Comment (GittiHab) [inline active_learning/active_learner.py]: For this class mostly the same comments apply as in the previous one.

File: active_learning/hypothesis.py [hunk 2]
```diff
@@ -30,3 +36,27 @@ def rank_meta_path(self, meta_path: MetaPath) -> float:
             return 1.0
         else:
             return 0.0
+
+
+class GaussianProcessHypothesis:
```
Comment (GittiHab) [inline active_learning/hypothesis.py]: Does this not have a common parent class with `Hypothesis`?

Comment (GittiHab) [inline active_learning/hypothesis.py]: Comment?

File: active_learning/hypothesis.py [hunk 3]
```diff
@@ -30,3 +36,27 @@ def rank_meta_path(self, meta_path: MetaPath) -> float:
...
+    def __init__(self, meta_paths):
```
Comment (GittiHab) [inline active_learning/hypothesis.py]: Parameter types...

File: active_learning/hypothesis.py [hunk 4]
```diff
@@ -30,3 +36,27 @@ def rank_meta_path(self, meta_path: MetaPath) -> float:
...
+        kernel = 1.0 * RBF(length_scale=1.0, length_scale_bounds=(1e-1, 10.0))
+        self.gp = GaussianProcessRegressor(kernel=kernel)
+        self.meta_paths = self._transform(meta_paths)
+
+    def _transform(self,meta_paths):
```
Comment (GittiHab) [inline active_learning/hypothesis.py]: Again, parameter and return types...

File: active_learning/hypothesis.py [hunk 5]
```diff
@@ -30,3 +36,27 @@ def rank_meta_path(self, meta_path: MetaPath) -> float:
...
+        """
+        Trivial transformation into feature space of length x unique_length.
+        """
+        return np.array([[len(mp), len(set(mp.as_list()))] for mp in meta_paths])
+
+    def update(self, idx, ratings):
```
Comment (GittiHab) [inline active_learning/hypothesis.py]: And again, parameter and return types...

File: active_learning/hypothesis.py [hunk 6]
```diff
@@ -30,3 +36,27 @@ def rank_meta_path(self, meta_path: MetaPath) -> float:
...
+        if len(idx) == 0:
+            return []
+        self.gp.fit(self.meta_paths[idx], ratings)
+
+    def predict_rating(self, idx):
```
Comment (GittiHab) [inline active_learning/hypothesis.py]: See previous method comments...

File: active_learning/hypothesis.py [hunk 7]
```diff
@@ -30,3 +36,27 @@ def rank_meta_path(self, meta_path: MetaPath) -> float:
...
+        return self.gp.predict(self.meta_paths[idx])
+
+    def predict_std(self, idx):
```
Comment (GittiHab) [inline active_learning/hypothesis.py]: Same as above...

File: active_learning/hypothesis.py [hunk 8]
```diff
@@ -30,3 +36,27 @@ def rank_meta_path(self, meta_path: MetaPath) -> float:
             return 1.0
         else:
             return 0.0
+
+
+class GaussianProcessHypothesis:
+    def __init__(self, meta_paths):
+        kernel = 1.0 * RBF(length_scale=1.0, length_scale_bounds=(1e-1, 10.0))
+        self.gp = GaussianProcessRegressor(kernel=kernel)
```
Comment (GittiHab) [inline active_learning/hypothesis.py]: `gp` is "very" descriptive...

File: api/server.py [hunk 9]
```diff
@@ -24,15 +24,14 @@
 app.config["SECRET_KEY"] = "37Y,=i9.,U3RxTx92@9j9Z[}"
 Session(app)
 
-if REACT_PORT is not 80:
-    CORS(app, supports_credentials=True, resources={r"/*": {"origins": "http://localhost:{}".format(REACT_PORT)}})
-else:
-    CORS(app, supports_credentials=True, resources={r"/*": {"origins": "http://localhost"}})
+#TODO revert
```
Comment (GittiHab) [inline api/server.py]: `TODO`

**Comment by GittiHab** [comment]
The coverage went down 😕 – maybe write some tests?

**Comment by feeds** [comment]
Hey @GittiHab, I apprechiate the effort you put into your feedback. 
Overall, I can agree on most of the points you make. 
On the other hand, I am refactring everything continously because I am still trying to figure out what the best architecture for our part of the framework is. This is why very neat programming is not exactly what I am doing here. It is rather scatching out and adapting to new ideas. Right now I am the only one using the stack, and while I am trying to leave something useful for @mutandon to work on, I can save time right now for trying ideas out. 
In todays standup we agreed to find a slot to talk about how software development vs. research friendly out master should be, i.e. what is acceptable to merge and what cannot be merged.

**Pull request merged**